### PR TITLE
Implement XWaitFor for Features

### PR DESCRIPTION
**Pull request by mateoflorido**
## Overview
Implement the `x-wait-for` commands for the DNS and Network components.

### CODE REVIEW

File: src/k8s/pkg/k8sd/features/cilium/status.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/canonical/k8s/pkg/snap"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func CheckNetwork(ctx context.Context, snap snap.Snap) error {
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/features/cilium/status.go:14]: This `CheckNetwork` should be a one-time check whether the pods are up and running. Do not add any retry mechanism

File: src/k8s/pkg/k8sd/features/coredns/status.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	"strings"
+	"time"
+
+	"github.com/canonical/k8s/pkg/snap"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CheckDNS checks the CoreDNS deployment in the cluster.
+func CheckDNS(ctx context.Context, snap snap.Snap) error {
```
Comment (neoaggelos) [inline src/k8s/pkg/k8sd/features/coredns/status.go:15]: Same here. `CheckDNS` should only check once whether the expected pods are up and running

File: src/k8s/cmd/k8s/k8s_x_wait_for_dns.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+	"github.com/spf13/cobra"
+)
+
+func newXWaitForDNSCmd(env cmdutil.ExecutionEnvironment) *cobra.Command {
+	waitForDNSCmd := &cobra.Command{
+		Use:    "x-wait-for-dns",
+		Short:  "Wait for DNS to be ready",
+		Hidden: true,
+		Run: func(cmd *cobra.Command, args []string) {
+			features.StatusChecks.CheckDNS(cmd.Context(), env.Snap)
```
Comment (neoaggelos) [inline src/k8s/cmd/k8s/k8s_x_wait_for_dns.go:15]: then, this command should have a timeout and use `control.WaitUntilReady`, calling CheckDNS repeatedly

File: src/k8s/cmd/k8s/k8s_x_wait_for_network.go [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+import (
+	cmdutil "github.com/canonical/k8s/cmd/util"
+	"github.com/canonical/k8s/pkg/k8sd/features"
+	"github.com/spf13/cobra"
+)
+
+func newXWaitForNetworkCmd(env cmdutil.ExecutionEnvironment) *cobra.Command {
+	waitForNetworkCmd := &cobra.Command{
+		Use:    "x-wait-for-network",
```
Comment (neoaggelos) [inline src/k8s/cmd/k8s/k8s_x_wait_for_network.go:11]: I'd make a top-level `x-wait-for` hidden command, and have `network`, `dns`, etc as (not hidden) sub-commands

**Review by neoaggelos — COMMENTED** [review]
Really nice work, left some comments that will help with de-duplication of code for alternative implementations

File: src/k8s/cmd/k8s/k8s_x_wait_for_dns.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+import (
+	cmdutil "github.com/canonical/k8s/cmd/util"
+	"github.com/canonical/k8s/pkg/k8sd/features"
+	"github.com/spf13/cobra"
+)
+
+func newXWaitForDNSCmd(env cmdutil.ExecutionEnvironment) *cobra.Command {
+	waitForDNSCmd := &cobra.Command{
+		Use:    "x-wait-for-dns",
```
Comment (louiseschmidtgen) [inline src/k8s/cmd/k8s/k8s_x_wait_for_dns.go:11]: Are we set on the name of this command? I would like to suggest something like feature-ready or feature-status?

File: src/k8s/pkg/k8sd/features/cilium/status.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	"context"
+	"fmt"
+
+	"github.com/canonical/k8s/pkg/snap"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func CheckNetwork(ctx context.Context, snap snap.Snap) (bool, error) {
+	client, err := snap.KubernetesClient("")
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/features/cilium/status.go:13]: nit: We only operate in the `kube-system` namespace in this function, hence the client namespace could be limited to `kube-system`

File: src/k8s/pkg/k8sd/features/coredns/status.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+	"fmt"
+
+	"github.com/canonical/k8s/pkg/snap"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CheckDNS checks the CoreDNS deployment in the cluster.
+func CheckDNS(ctx context.Context, snap snap.Snap) (bool, error) {
+	client, err := snap.KubernetesClient("")
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/features/coredns/status.go:14]: same here

**Review by bschimke95 — COMMENTED** [review]
Great work @mateoflorido, only minor comments.

I think it makes sense to include the `wait_for_x` changes for the Python tests into this PR as well. That way, we have all changes for this refactoring in one PR, WDYT?

File: src/k8s/cmd/k8s/k8s_x_wait_for_dns.go (hunk 5 above)
Comment (bschimke95) [reply src/k8s/cmd/k8s/k8s_x_wait_for_dns.go]: I prefer the current naming. `dns-ready` or `dns-status` sound like querying a state while `wait-for-dns` clearly shows that this is a blocking invocation.

Comment (mateoflorido) [reply src/k8s/cmd/k8s/k8s_x_wait_for_dns.go]: I agree with Ben. This naming convention is more self-descriptive of the command's purpose.

**Comment by mateoflorido** [comment]
@bschimke95 I've changed the utils in the integration tests to use the XWaitFor commands. I've also scoped the client to the `kube-system` namespace. Thanks for the review!

**Review by louiseschmidtgen — APPROVED** [review]
LGTM, thanks Mateo!

**Pull request merged**